File: backend/src/Controllers/PromptLibraryController.php
<?php

declare(strict_types=1);

namespace Quantis\AIPortfolioAssistant\Controllers;

use PDO;
use Exception;

/**
 * Prompt Library Controller
 *
 * Manages CRUD operations for hierarchical prompt library.
 */
class PromptLibraryController
{
    private PDO $db;
    private array $config;

    public function __construct(PDO $db, array $config)
    {
        $this->db = $db;
        $this->config = $config;
    }

    /**
     * Get prompt library tree structure for user
     */
    public function getTree(array $request): array
    {
        $userId = $request['user_id'];

        // Fetch all items for user
        $stmt = $this->db->prepare('
            SELECT
                id,
                parent_id,
                type,
                name,
                content,
                sort_order,
                created_at,
                updated_at
            FROM prompt_library
            WHERE user_id = ?
            ORDER BY parent_id, sort_order, name
        ');
        $stmt->execute([$userId]);
        $items = $stmt->fetchAll();

        // Build hierarchical tree
        $tree = $this->buildTree($items);

        return [
            'success' => true,
            'data' => $tree,
            'status_code' => 200
        ];
    }

    /**
     * Get a specific prompt library item by ID
     */
    public function get(array $request, int $id): array
    {
        $userId = $request['user_id'];

        $stmt = $this->db->prepare('
            SELECT
                id,
                parent_id,
                type,
                name,
                content,
                sort_order,
                created_at,
                updated_at
            FROM prompt_library
            WHERE id = ? AND user_id = ?
        ');
        $stmt->execute([$id, $userId]);
        $item = $stmt->fetch();

        if (!$item) {
            return [
                'success' => false,
                'message' => 'Item not found',
                'status_code' => 404
            ];
        }

        return [
            'success' => true,
            'data' => $item,
            'status_code' => 200
        ];
    }

    /**
     * Create new prompt or folder
     */
    public function create(array $request): array
    {
        $userId = $request['user_id'];
        $input = $request['body'];

        // Validate required fields
        if (empty($input['type']) || !in_array($input['type'], ['folder', 'prompt'])) {
            return [
                'success' => false,
                'message' => 'Valid type (folder or prompt) is required',
                'status_code' => 400
            ];
        }

        if (empty($input['name'])) {
            return [
                'success' => false,
                'message' => 'Name is required',
                'status_code' => 400
            ];
        }

        // Extract data
        $type = $input['type'];
        $name = trim($input['name']);
        $parentId = $input['parent_id'] ?? null;
        $content = ($type === 'prompt' && isset($input['content'])) ? $input['content'] : null;
        $sortOrder = $input['sort_order'] ?? 0;

        // Validate parent exists if specified
        if ($parentId !== null) {
            $stmt = $this->db->prepare('SELECT id FROM prompt_library WHERE id = ? AND user_id = ?');
            $stmt->execute([$parentId, $userId]);
            if (!$stmt->fetch()) {
                return [
                    'success' => false,
                    'message' => 'Parent folder not found',
                    'status_code' => 400
                ];
            }
        }

        // Insert new item
        $stmt = $this->db->prepare('
            INSERT INTO prompt_library (user_id, parent_id, type, name, content, sort_order)
            VALUES (?, ?, ?, ?, ?, ?)
        ');
        $stmt->execute([$userId, $parentId, $type, $name, $content, $sortOrder]);
        $newId = $this->db->lastInsertId();

        return [
            'success' => true,
            'message' => ucfirst($type) . ' created successfully',
            'data' => [
                'id' => $newId,
                'type' => $type,
                'name' => $name,
                'parent_id' => $parentId,
                'content' => $content
            ],
            'status_code' => 200
        ];
    }

    /**
     * Update existing prompt or folder
     */
    public function update(array $request, int $id): array
    {
        $userId = $request['user_id'];
        $input = $request['body'];

        // Check if item exists and belongs to user
        $stmt = $this->db->prepare('SELECT id, type FROM prompt_library WHERE id = ? AND user_id = ?');
        $stmt->execute([$id, $userId]);
        $item = $stmt->fetch();

        if (!$item) {
            return [
                'success' => false,
                'message' => 'Item not found',
                'status_code' => 404
            ];
        }

        // Build update query dynamically based on provided fields
        $updates = [];
        $params = [];

        if (isset($input['name']) && !empty(trim($input['name']))) {
            $updates[] = 'name = ?';
            $params[] = trim($input['name']);
        }

        if (isset($input['content'])) {
            $updates[] = 'content = ?';
            $params[] = $input['content'];
        }

        if (isset($input['parent_id'])) {
            $updates[] = 'parent_id = ?';
            $params[] = $input['parent_id'];
        }

        if (isset($input['sort_order'])) {
            $updates[] = 'sort_order = ?';
            $params[] = $input['sort_order'];
        }

        if (empty($updates)) {
            return [
                'success' => false,
                'message' => 'No fields to update',
                'status_code' => 400
            ];
        }

        // Add WHERE parameters
        $params[] = $id;
        $params[] = $userId;

        // Execute update
        $sql = 'UPDATE prompt_library SET ' . implode(', ', $updates) . ' WHERE id = ? AND user_id = ?';
        $stmt = $this->db->prepare($sql);
        $stmt->execute($params);

        return [
            'success' => true,
            'message' => ucfirst($item['type']) . ' updated successfully',
            'data' => ['id' => $id],
            'status_code' => 200
        ];
    }

    /**
     * Delete prompt or folder (cascades to children)
     */
    public function delete(array $request, int $id): array
    {
        $userId = $request['user_id'];

        // Check if item exists and belongs to user
        $stmt = $this->db->prepare('SELECT id, type, name FROM prompt_library WHERE id = ? AND user_id = ?');
        $stmt->execute([$id, $userId]);
        $item = $stmt->fetch();

        if (!$item) {
            return [
                'success' => false,
                'message' => 'Item not found',
                'status_code' => 404
            ];
        }

        // Delete item (CASCADE will handle children)
        $stmt = $this->db->prepare('DELETE FROM prompt_library WHERE id = ? AND user_id = ?');
        $stmt->execute([$id, $userId]);

        return [
            'success' => true,
            'message' => ucfirst($item['type']) . ' deleted successfully',
            'status_code' => 200
        ];
    }

    /**
     * Build hierarchical tree from flat array
     */
    private function buildTree(array $items, ?int $parentId = null): array
    {
        $branch = [];

        foreach ($items as $item) {
            if ($item['parent_id'] == $parentId) {
                $children = $this->buildTree($items, (int)$item['id']);
                if ($children) {
                    $item['children'] = $children;
                }
                $branch[] = $item;
            }
        }

        return $branch;
    }
}

File: backend_python/app/controllers/prompt_library_controller.py
"""Port of Controllers/PromptLibraryController.php."""
from __future__ import annotations

from app.support.phpcompat import php_empty, ucfirst

_SELECT = ('SELECT id, parent_id, type, name, content, sort_order, created_at, updated_at'
           ' FROM prompt_library')


def _loose_eq(a, b) -> bool:
    """PHP `==` for the parent_id comparison: null==null, "3"==3, 3==3."""
    if a is None or b is None:
        return a is None and b is None
    try:
        return int(a) == int(b)
    except (TypeError, ValueError):
        return str(a) == str(b)


def build_tree(items: list[dict], parent_id=None) -> list[dict]:
    branch = []
    for item in items:
        if _loose_eq(item.get('parent_id'), parent_id):
            node = dict(item)
            children = build_tree(items, int(item['id']))
            if children:
                node['children'] = children
            branch.append(node)
    return branch


class PromptLibraryController:
    def __init__(self, db, config):
        self.db = db
        self.config = config

    def getTree(self, request) -> dict:
        items = self.db.fetch_all(_SELECT + ' WHERE user_id = ? ORDER BY parent_id, sort_order, name',
                                  [request['user_id']])
        return {'success': True, 'data': build_tree(items), 'status_code': 200}

    def get(self, request, id: int) -> dict:
        item = self.db.fetch_one(_SELECT + ' WHERE id = ? AND user_id = ?', [id, request['user_id']])
        if not item:
            return {'success': False, 'message': 'Item not found', 'status_code': 404}
        return {'success': True, 'data': item, 'status_code': 200}

    def create(self, request) -> dict:
        user_id = request['user_id']
        inp = request['body']
        if not inp.get('type') or inp['type'] not in ('folder', 'prompt'):
            return {'success': False, 'message': 'Valid type (folder or prompt) is required', 'status_code': 400}
        if php_empty(inp.get('name')):
            return {'success': False, 'message': 'Name is required', 'status_code': 400}
        type_ = inp['type']
        name = str(inp['name']).strip()
        parent_id = inp.get('parent_id')
        content = inp['content'] if (type_ == 'prompt' and 'content' in inp) else None
        sort_order = inp.get('sort_order')
        if sort_order is None:
            sort_order = 0
        if parent_id is not None:
            if not self.db.fetch_one('SELECT id FROM prompt_library WHERE id = ? AND user_id = ?', [parent_id, user_id]):
                return {'success': False, 'message': 'Parent folder not found', 'status_code': 400}
        new_id = self.db.insert(
            'INSERT INTO prompt_library (user_id, parent_id, type, name, content, sort_order) VALUES (?, ?, ?, ?, ?, ?)',
            [user_id, parent_id, type_, name, content, sort_order])
        return {'success': True, 'message': ucfirst(type_) + ' created successfully',
                'data': {'id': str(new_id), 'type': type_, 'name': name, 'parent_id': parent_id, 'content': content},
                'status_code': 200}

    def update(self, request, id: int) -> dict:
        user_id = request['user_id']
        inp = request['body']
        item = self.db.fetch_one('SELECT id, type FROM prompt_library WHERE id = ? AND user_id = ?', [id, user_id])
        if not item:
            return {'success': False, 'message': 'Item not found', 'status_code': 404}
        updates, params = [], []
        if inp.get('name') is not None and str(inp['name']).strip():
            updates.append('name = ?'); params.append(str(inp['name']).strip())
        if inp.get('content') is not None:
            updates.append('content = ?'); params.append(inp['content'])
        if inp.get('parent_id') is not None:
            updates.append('parent_id = ?'); params.append(inp['parent_id'])
        if inp.get('sort_order') is not None:
            updates.append('sort_order = ?'); params.append(inp['sort_order'])
        if not updates:
            return {'success': False, 'message': 'No fields to update', 'status_code': 400}
        params += [id, user_id]
        self.db.execute('UPDATE prompt_library SET ' + ', '.join(updates) + ' WHERE id = ? AND user_id = ?', params)
        return {'success': True, 'message': ucfirst(item['type']) + ' updated successfully', 'data': {'id': id},
                'status_code': 200}

    def delete(self, request, id: int) -> dict:
        user_id = request['user_id']
        item = self.db.fetch_one('SELECT id, type, name FROM prompt_library WHERE id = ? AND user_id = ?', [id, user_id])
        if not item:
            return {'success': False, 'message': 'Item not found', 'status_code': 404}
        self.db.execute('DELETE FROM prompt_library WHERE id = ? AND user_id = ?', [id, user_id])
        return {'success': True, 'message': ucfirst(item['type']) + ' deleted successfully', 'status_code': 200}
